### PR TITLE
Allow BabeConsensusDataProvider fork existing chain

### PR DESCRIPTION
**Pull request by seunlanlege**
This allows `BabeConsensusDataProvider` build blocks on top of an existing chain where the Alice key isn't part of the authorities.

### CODE REVIEW

File: client/consensus/babe/src/lib.rs
Comment (tomusdrw) [inline client/consensus/babe/src/lib.rs:202]: Should rather have docs.

File: client/consensus/babe/src/aux_schema.rs
Comment (tomusdrw) [inline client/consensus/babe/src/aux_schema.rs:31]:
```suggestion
/// Database key for epoch changes version (AUX column).
```

File: client/consensus/babe/src/aux_schema.rs
Comment (tomusdrw) [inline client/consensus/babe/src/aux_schema.rs:33]:
```suggestion
/// Database key for epoch changes (AUX column).
```

File: client/consensus/manual-seal/src/consensus/babe.rs
Comment (tomusdrw) [inline client/consensus/manual-seal/src/consensus/babe.rs:134]: Why do we explicitly drop it?

File: client/consensus/manual-seal/src/consensus/babe.rs
Comment (tomusdrw) [inline client/consensus/manual-seal/src/consensus/babe.rs:189]: This (+ the context) seems like a duplicate of what is in `create_digest`, any way to DRY it?

File: client/consensus/manual-seal/src/consensus/babe.rs
Comment (tomusdrw) [inline client/consensus/manual-seal/src/consensus/babe.rs:153]: I was actually hoping that we can pretend to be regular babe authorities instead of using Alice's key. Does this somehow leak to runtime? I.e. imagine that in the testing environment we would probably want to test that if one of the validators misbehaves Babe starts ignoring other blocks produced by that validator.
Is it required that this here is always `Alice`? Could we make this configurable?

File: client/consensus/manual-seal/src/consensus/babe.rs
Comment (tomusdrw) [inline client/consensus/manual-seal/src/consensus/babe.rs:209]: Why do we use `babe` target? Shouldn't that be `manual-seal`?

File: client/consensus/manual-seal/src/consensus/babe.rs
Comment (tomusdrw) [inline client/consensus/manual-seal/src/consensus/babe.rs:204]: Same here. For tests it would probably be useful to be able to configure authorities.

**Review by tomusdrw — COMMENTED** [review]
Looks reasonable. Again the context here is to be able to create "fake" babe blocks in the test environment with `ManualSeal`.

File: client/consensus/manual-seal/src/consensus/babe.rs
Comment (seunlanlege) [reply client/consensus/manual-seal/src/consensus/babe.rs:134]: the epoch holds a ref to shared_epoch_changes, which prevents us from acquiring a mutable reference to epoch in line 146

File: client/consensus/manual-seal/src/consensus/babe.rs
Comment (seunlanlege) [reply client/consensus/manual-seal/src/consensus/babe.rs:209]: yeah, will fix it

File: client/consensus/manual-seal/src/consensus/babe.rs
Comment (seunlanlege) [reply client/consensus/manual-seal/src/consensus/babe.rs:153]: > I was actually hoping that we can pretend to be regular babe authorities instead of using Alice's key.

hmm, actually now that I think about it, we could probably just use secondary slots and never have to mutate the epoch. Will have to test this. But for now, will do this: 

> Could we make this configurable?

File: primitives/io/src/lib.rs
Comment (bkchr) [inline primitives/io/src/lib.rs:413]: Ideally this should not live in Substrate.

If it needs to live in Substrate, it should go into a different crate. Where it is directly understandable that this is for testing.

File: primitives/io/src/lib.rs
Comment (tomusdrw) [inline primitives/io/src/lib.rs:501]: Please change the state of this PR to Draft since it seems it's being used to experiment with stuff and for sure is not ready to be merged to Substrate.

**Review by tomusdrw — CHANGES_REQUESTED** [review]

File: client/api/src/execution_extensions.rs
Comment (tomusdrw) [inline client/api/src/execution_extensions.rs:102]: I think it shouldn't be required since it's going to be rarely customized. We should either add it not as an extension but rather to `Externalities` trait (with default sane implementations) or we should allow overriding via `extensions_factory` - i.e. we register the default extension only if it's not present in the `Extensions` returned by `extensions_factory` invocation.

File: primitives/core/src/traits.rs
Comment (tomusdrw) [inline primitives/core/src/traits.rs:237]: To prevent a virtual call in the default case, we could perhaps do:
```
enum CryptoExt {
  Default(DefaultCryptoImpl),
  Custom(Box<dyn CryptoExt>),
}
```

Alternatively we could change the method implementation so that in case `CryptoExt` is not there we fall back to the default impl (i.e. `CryptoExtension` would not be required to be registered for this thing to work), but I feel current approach feels cleaner.

**Review by tomusdrw — APPROVED** [review]
looks good to me, @bkchr could you take a look? I remember you didn't like the idea of putting crypto stuff into an extension, but we couldn't make it work with `replace_implementation`, since that one has to be called from within the runtime (and we want to keep the existing runtime and only alter the host environment).

**Comment by bkchr** [comment]
> looks good to me, @bkchr could you take a look? I remember you didn't like the idea of putting crypto stuff into an extension, but we couldn't make it work with `replace_implementation`, since that one has to be called from within the runtime (and we want to keep the existing runtime and only alter the host environment).

I never said that @seunlanlege should use `replace_implementation`. I described it to him, that the host function that comes first is chosen and I also had shown him where he needs to change the order to make this work...

File: client/api/src/execution_extensions.rs
Comment (bkchr) [inline client/api/src/execution_extensions.rs:171]:
```suggestion
		if extensions.get_mut(TypeId::of::<CryptoExtension>()).is_none() {
```

File: primitives/core/src/traits.rs
Comment (bkchr) [inline primitives/core/src/traits.rs:211]: This function is only required for Kusama to sync. We don't require it here.

File: primitives/core/src/traits.rs
Comment (bkchr) [inline primitives/core/src/traits.rs:198]: Docs are not correct.

File: primitives/core/src/traits.rs
Comment (bkchr) [inline primitives/core/src/traits.rs:199]: What is with the batching stuff, why isn't that here? Where do we draw the line?

**Comment by bkchr** [comment]
@NikVolf is your benchmark bot still working? Could you please run it here.

File: primitives/core/src/traits.rs
Comment (tomusdrw) [reply primitives/core/src/traits.rs:199]: Good catch, we should defintely move all crypto stuff here to make it consistent.

Comment (seunlanlege) [reply primitives/core/src/traits.rs:199]: batching eventually calls methods on `CryptoExt`

**Comment by bkchr** [comment]
@seunlanlege please explain me what the problem was with overwriting the host functions?

**Comment by seunlanlege** [comment]
Asides from the fact that it didn't work for me for some unknown reason, We needed a way to override signature verification without forking substrate and modifying the order in which host functions are initialized.

**Comment by bkchr** [comment]
> Asides from the fact that it didn't work for me for some unknown reason, 

That is no "explanation" I will accept here.

>We needed a way to override signature verification without forking substrate and modifying the order in which host functions are initialized.

You already have done the changes here: https://github.com/paritytech/substrate/pull/7078/files#diff-b447ad296633d7d693790a1fb7ce8b5f ?

And you modify Substrate here anyway?

**Review by tomusdrw — CHANGES_REQUESTED** [review]
Seems that crypto overrides can be better done with a host function overrides than execution extension, so let's implement that instead.

File: frame/system/src/lib.rs
Comment (bkchr) [inline frame/system/src/lib.rs:427]:
```suggestion
		Events get(fn events): Vec<EventRecord<T::Event, T::Hash>>;
```

`system::Module::events()` is what you want

File: client/api/src/execution_extensions.rs
Comment (bkchr) [inline client/api/src/execution_extensions.rs:141]:
```suggestion
	pub fn extensions(&self, at: &BlockId<Block>, context: ExecutionContext) -> Extensions {
```

**Comment by tomusdrw** [comment]
@seunlanlege this PR doesn't have a single approval, can you justify why it was merged?

**Comment by seunlanlege** [comment]
> this PR doesn't have a single approval

weird, could've sworn @bkchr approved it, which was why i was able to merge

**Comment by bkchr** [comment]
> > this PR doesn't have a single approval
> 
> weird, could've sworn @bkchr approved it, which was why i was able to merge

I did not approve anything here.

**Comment by seunlanlege** [comment]
Huh, merge button was green for some other reason then.

**Comment by bkchr** [comment]
Yeah, but that is no excuse for merging this.

**Comment by seunlanlege** [comment]
Aren't approvals are a requirement for merging?

**Pull request merged**
